### ISSUE TITLE
DomainObjectCollection is not thread-safe and can cause an infinite loop

### ISSUE THREAD

**Issue by wilkinsona** [gradle/gradle#6796]
### Expected Behavior

Plugins and build scripts can mutate a `DomainObjectCollection` without having to worry about thread-safety.

### Current Behavior

Parallel threads mutating the same `DomainObjectCollection` can corrupt the underlying `TreeMap` causing reads and write to loop infinitely.

### Context

https://github.com/ben-manes/gradle-versions-plugin/issues/255

There's a reasonably complex interaction here between the Versions Plugin and the Dependency Management Plugin. The former is now performing some configuration resolution on multiple threads in parallel. This causes the dependency management plugin to add items to a `DomainObjectCollection` on multiple threads in parallel. This causes the above-described corruption of the underlying `TreeMap`.

### Steps to Reproduce (for bugs)

Here's a, somewhat contrived, build script that should reproduce the problem:

```groovy
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

def pool = Executors.newFixedThreadPool(10)
def count = new AtomicInteger()
def latch = new CountDownLatch(1000)
for (int i = 0; i < 10; i++) {
        pool.submit {
                for (int j = 0; j < 100; j++) {
                        def name = "test${count.incrementAndGet()}"
                        println name
                        project.configurations.create(name)
                        latch.countDown()
                }
        }
}
latch.await()
```

There's also a sample project in the issue referenced above that reproduces the problem in a real-world scenario.

### Your Environment

```
------------------------------------------------------------
Gradle 4.10.1
------------------------------------------------------------

Build time:   2018-09-12 11:33:27 UTC
Revision:     76c9179ea9bddc32810f9125ad97c3315c544919

Kotlin DSL:   1.0-rc-6
Kotlin:       1.2.61
Groovy:       2.4.15
Ant:          Apache Ant(TM) version 1.9.11 compiled on March 23 2018
JVM:          1.8.0_181 (Oracle Corporation 25.181-b13)
OS:           Mac OS X 10.13.6 x86_64
```

**Comment by big-guy** [comment]
`DomainObjectCollection`s aren't intended to be thread safe. 

> There's a reasonably complex interaction here between the Versions Plugin and the Dependency Management Plugin. The former is now performing some configuration resolution on multiple threads in parallel. 

This shouldn't be necessary. Gradle is already going to be resolving things in parallel and this will fail in very strange ways (hanging, NPE, concurrent modification exceptions, etc).  

This doesn't mean you should sprinkle synchronization into the dependency management plugin.  I think you should be able to assume that you don't need to deal with concurrent access like this, but we're not going to make domain object collection thread-safe.

I'll reply to the other issue.

**Comment by ben-manes** [comment]
In the Versions plugin, we resolve configurations in parallel as the metadata queries (pom lookups) are not parallelized by Gradle, yet. The artifact jar is downloaded in parallel, though, so that covers the common-case (but doesn't help us). We used GPars to parallelize this since it sounded thread-safe now and, in our case, do read only logic to resolve.

Some ugly code to invoke the resolution is:
```groovy
def pool = GParsPool.createPool(numberOfThreads)
return GParsPool.withExistingPool(pool) {
  projectConfigs.keySet().collectParallel { proj ->
    Set<Configuration> configurations = projectConfigs.get(proj)
    Resolver resolver = new Resolver(proj, resolutionStrategy, pool)
    GParsPool.withExistingPool(pool) {
      configurations.collectParallel { Configuration config ->
        resolve(resolver, proj, config)
      }.flatten() as Set<DependencyStatus>
    }
  }.flatten() as Set<DependencyStatus>
}
```

Each configuration is processed in parallel, as it should be independent. In simplified form, we're invoking `configuration.resolvedConfiguration.lenientConfiguration.getFirstLevelModuleDependencies()` with the original set and a clone using `+` for the version string. The aspects are aggregated for the eventual report output.

It's not clear from an API perspective if this is expected to be thread-safe or not, since partial parallel resolution support was added to Gradle and not done for the `pom` downloads as low priority work.

**Comment by wilkinsona** [comment]
> It's not clear from an API perspective if this is expected to be thread-safe or not.

This is the key point from my perspective. I think it would be very beneficial to plugin developers if Gradle's documentation discussed thread-safety. That could be something in the reference documentation, a guide, or the javadoc of each class, or, ideally, some combination of the three. As things stand, I find it very difficult to know what thread-safety guarantees a plugin can expect.

**Comment by oehme** [comment]
Resolving Configurations in parallel should be thread safe. I assume there is some missing synchronization around user-provided code (e.g. `beforeResolve`). We should fix those.

**Comment by ldaley** [comment]
We should not lose @wilkinsona's suggestion to actually explain the concurrency assumptions and constructs in Gradle to logic authors, and the core developers.
